accounting/accounting_test.go

```diff
@@ -0,0 +1,1328 @@
+				}
+			}
+		})
+	}
+}
+
+func TestRatNetAmount(t *testing.T) {
+	br := func(f64 float64) *big.Rat {
+		return new(big.Rat).SetFloat64(f64)
+	}
+
+	type args struct {
+		value *big.Rat
+		rate  *big.Rat
+	}
+	var tests = []struct {
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "gross 10 vat 19",
+			args: args{
+				value: br(10),
+				rate:  br(.19),
+			},
+			want:    "8.403361345",
+			wantErr: false,
+		},
+		{
+			name: "gross 19.99 vat 20",
+			args: args{
+				value: br(19.99),
+				rate:  br(.20),
+			},
+			want:    "16.658333333",
+			wantErr: false,
+		},
+		{
+			name: "gross 123 vat 7",
+			args: args{
+				value: br(123),
+				rate:  br(.07),
+			},
+			want:    "114.953271028",
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := accounting.RatNetAmount(tt.args.value, tt.args.rate)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("RatNetAmount() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !cmp.Equal(got.Text('f', 9), tt.want) {
+				t.Error(cmp.Diff(got.Text('f', 9), tt.want))
+			}
+		})
+	}
+}
+
+func TestToMinorUnit(t *testing.T) {
+	for _, tt := range minorUnitTestsCases {
+		t.Run(tt.name, func(t *testing.T) {
+			i := accounting.ToMinorUnit(currency.GBP, tt.f64)
+			if !cmp.Equal(i, tt.i64) {
+				t.Error(cmp.Diff(i, tt.i64))
+			}
+		})
+	}
+}
+
+func TestFromMinorUnit(t *testing.T) {
+	for _, tt := range minorUnitTestsCases {
+		t.Run(tt.name, func(t *testing.T) {
+			f := accounting.FromMinorUnit(currency.GBP, tt.i64)
+			if !cmp.Equal(f, tt.f64) {
+				t.Error(cmp.Diff(f, tt.f64))
+			}
+		})
+	}
+}
+
+func TestValidateFloatIsPrecise(t *testing.T) {
+	tests := []struct {
+		name      string
+		amount    float64
+		wantError bool
+	}{
+		{
+			name:      "exact number",
+			amount:    10,
+			wantError: false,
+		},
+		{
+			name:      "too long",
+			amount:    12.123,
+			wantError: true,
+		},
+		{
+			name:      "valid amount",
+			amount:    12.12,
+			wantError: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := accounting.ValidateFloatIsPrecise(tt.amount)
+			if err != nil && !tt.wantError {
+				t.Fatalf(
+					"ValidateFloatIsPrecise: failed on %v",
+					tt.amount,
+				)
+			}
+		})
+	}
+}
+
+func TestValidateManyFloatsArePrecise(t *testing.T) {
+	tests := []struct {
+		name      string
+		amounts   []float64
+		wantError bool
+	}{
+		{
+			name:      "exact numbers",
+			amounts:   []float64{10, 11, 12},
+			wantError: false,
+		},
+		{
+			name:      "one is too long",
+			amounts:   []float64{11, 12.123, 13},
+			wantError: true,
+		},
+		{
+			name:      "valid amounts",
+			amounts:   []float64{11.11, 12.12, 13.13},
+			wantError: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := accounting.ValidateManyFloatsArePrecise(tt.amounts...)
+			if err != nil && !tt.wantError {
```

Comment:
same as above - this won't test all cases

Reply:
fixed